### PR TITLE
Add support for l2, minmax, zscore normalization

### LINKED ISSUE #37: Add option to specify normalization method
When normalizing sentence transformer embeddings, we currently only support L2 normalization or no normalization. As @jteijema pointed out, it would be good to experiment with other normalizations (such as minmax), since classification is a different use case than what L2 normalization is usually ideal for.

The proposal is to add an additional parameter `normalization_method` to the sentence transformer base class, with a default of L2 normalization as to keep compatibility with ASReview Core.

### PR DESCRIPTION
**Pull request by timovdk**
Closes #37

### CODE REVIEW

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py [hunk 1]
```diff
@@ -98,16 +141,40 @@ def fit_transform(self, X, y=None):
...
+        if self.normalize_method == "l2":
+            return sklearn_normalize(embeddings, norm="l2")
+        elif self.normalize_method == "minmax":
+            return MinMaxScaler().fit_transform(embeddings)
+        elif self.normalize_method == "standard":
+            return StandardScaler().fit_transform(embeddings)
+        else:
+            raise ValueError(
+                f"Unsupported normalization method: '{self.normalize_method}'"
+            )
```
Comment (J535D165) [inline asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:176]: Using the pipeline would be nicer and more in line with the API design

```python
        steps = [
            ("text_merger", TextMerger(columns=self.columns, sep=self.sep)),
            (
                "sentence_transformer",
                BaseSentenceTransformer(
                    model_name=self.model_name,
                    normalize=self.normalize,
                    normalize_method=self.normalize_method,
                    quantize=self.quantize,
                    precision=self.precision,
                    verbose=self.verbose,
                ),
            ),
        ]
        super().__init__(steps)
```

**Review by J535D165 — COMMENTED** [review]
Nice addition. Curious about the result.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 1 above)
Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:176]: What do you mean exactly? To pass a callable from `SentenceTransformerPipeline` to `BaseSentenceTranformer` as normalize_method, so that this if/else logic moves to `SentenceTransformerPipeline`?

Comment (J535D165) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:176]: Something like this in `SentenceTransformerPipeline`:

```python
        steps = [
            ("text_merger", TextMerger(columns=self.columns, sep=self.sep)),
            (
                "sentence_transformer",
                BaseSentenceTransformer(
                    model_name=self.model_name,
                    normalize=self.normalize,
                    normalize_method=self.normalize_method,
                    quantize=self.quantize,
                    precision=self.precision,
                    verbose=self.verbose,
                ),
            ),
        ]
        
        if self.normalize_method == "minmax":
            steps.append(MinMaxScaler())
        
        if self.normalize_method == "standard":
             steps.append(StandardScaler())
        
        super().__init__(steps)

```

I don't understand L2 completely, as this seems to be the default as well, isn't it? So I think you can leave that one out. 

I think this would be more in line with Keras and sklearn API design principles.

Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:176]: Oh yeah that's nice, thanks!

Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py]: @J535D165 To make this work, quantization should also be a pipeline step (which actually follows the structure better as well), because quantization cannot happen before normalization.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py [hunk 2]
```diff
@@ -94,20 +147,23 @@ def fit(self, X, y=None):
         # for sentence-transformers, so return self
         return self
 
-    def fit_transform(self, X, y=None):
+    def transform(self, X, y=None):
         if self.verbose:
             print("Embedding text...")
 
-        embeddings = self._model.encode(
-            X, show_progress_bar=self.verbose, normalize_embeddings=self.normalize
```
Comment (J535D165) [inline asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:102]: I wonder what the `normalize_embedding` does when set to `True`. Just curiosity.

Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:102]: It normalizes using torch nn: `embeddings = torch.nn.functional.normalize(embeddings, p=2, dim=1)`

Comment (J535D165) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:102]: Oke, that's L2.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py [hunk 3]
```diff
@@ -94,20 +147,23 @@ def fit(self, X, y=None):
...
-                embeddings = embeddings.numpy()
         return embeddings
 
+    def _to_numpy(self, arr):
+        """Ensure input is a NumPy array."""
+        if hasattr(arr, "numpy"):
+            return arr.numpy()
+        if not isinstance(arr, np.ndarray):
+            return np.array(arr)
+        return arr
```
Comment (timovdk) [inline asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:165]: @J535D165 This function I do have some doubt about, whether it is actually necessary haha

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py [hunk 4]
```diff
@@ -46,13 +94,24 @@ def __init__(
...
                     model_name=self.model_name,
-                    normalize=self.normalize,
-                    quantize=self.quantize,
-                    precision=self.precision,
                     verbose=self.verbose,
                 ),
             ),
         ]
+        if self.normalize:
+            if self.normalize_method == "l2":
```
Comment (J535D165) [inline asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:102]: I like the overall design now, except for this. In general, I would expect [None, l2, minmax, standard]. In that case, you don't need two variables to achieve the same thing. But maybe you do this for backward compat.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 3 above)
Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:165]: Actually, I think we do need it, since there are 6 encode overloads that all return different datatypes. Sorry for the ping!

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 4 above)
Comment (timovdk) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:102]: Yes indeed, it's to not have to change the h3/l2 models (normalize==True with normalize_method's default is the same as it was). I can also make a PR to change h3/l2's parameters, I would prefer that tbh, to keep the API of Dory neat.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py [hunk 5]
```diff
@@ -46,13 +94,24 @@ def __init__(
                 "sentence_transformer",
                 BaseSentenceTransformer(
                     model_name=self.model_name,
-                    normalize=self.normalize,
```
Comment (J535D165) [inline asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:49]: If you do something like:

```python
normalize=self.normalize is True or self.normalize == "l2"
```

(The True ensures backward compat)

then you can simplify the rest of the code:

```python
if self.normalize == "minmax":
    steps.append(("normalizer", MinMaxScaler()))
elif self.normalize == "standard":
    steps.append(("normalizer", StandardScaler()))
```

you don't need the l2 from sklearn then anymore.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 1 above)
Comment (J535D165) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py]: Makes sense. I like your implementation

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 3 above)
Comment (J535D165) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:165]:
```python
hasattr(arr, "numpy"):
    return arr.numpy()
```
This part is, of course, very ugly. What kind of library implements this, haha

```python
        if not isinstance(arr, np.ndarray):
            return np.array(arr)
        return arr
```

Don't be too nice. `return np.array(arr)` is enough.

File: asreviewcontrib/dory/feature_extractors/transformer_embeddings.py (hunk 5 above)
Comment (J535D165) [reply asreviewcontrib/dory/feature_extractors/transformer_embeddings.py:49]: Oke! Good to me.

**Review by J535D165 — APPROVED** [review]

**Comment by J535D165** [comment]
Shall we make a 1.1 release?
